activerecord/lib/active_record/testing/query_assertions.rb

```diff
@@ -7,28 +7,102 @@ module QueryAssertions
       #
       #   assert_queries(1) { Post.first }
       #
-      # If the +:matcher+ option is provided, only queries that match the matcher are counted.
-      #
-      #   assert_queries(1, matcher: /LIMIT \?/) { Post.first }
+      # You can use +assert_sql+ assertion method for more advanced checks.
       #
-      def assert_queries(expected_count, matcher: nil, &block)
+      def assert_queries(expected_count, &block)
         ActiveRecord::Base.connection.materialize_transactions
 
-        queries = []
-        callback = lambda do |*, payload|
-          queries << payload[:sql] if %w[ SCHEMA TRANSACTION ].exclude?(payload[:name]) && (matcher.nil? || payload[:sql].match(matcher))
-        end
-        ActiveSupport::Notifications.subscribed(callback, "sql.active_record") do
+        counter = SQLCounter.new
+        ActiveSupport::Notifications.subscribed(counter, "sql.active_record") do
           result = _assert_nothing_raised_or_warn("assert_queries", &block)
+          queries = counter.log
           assert_equal expected_count, queries.size, "#{queries.size} instead of #{expected_count} queries were executed. Queries: #{queries.join("\n\n")}"
           result
         end
       end
 
       # Asserts that no SQL queries are executed in the given block.
+      #
+      #   assert_no_queries { post.comments }
+      #
       def assert_no_queries(&block)
         assert_queries(0, &block)
       end
+
+      # Asserts that the SQL queries executed in the given block match expected conditions.
+      #
+      #   # Check for exact number of queries
+      #   assert_sql(1) { Post.first }
+      #
+      #   # Check for any number of queries
+      #   assert_sql { Post.first }
+      #
+      # If the +:match+ option is provided, only queries that match the matcher are considered.
+      #
+      #   assert_sql(1, match: /LIMIT \?/) { Post.first }
+      #
+      # If the +:include_all+ option is provided, any queries (including transaction and schema related)
+      #   that match the matcher are considered.
+      #
+      #   assert_sql(include_all: true) { Post.columns }
+      #
+      def assert_sql(expected_count = nil, match: nil, include_all: false, &block)
```

Comment:
I think I'd inverse the parameters:

```suggestion
      def assert_sql(match, count: nil, include_all: false, &block)
```

What do you think?

Reply:
I originally thought for `match` to not be required to be able to write something like `assert_sql(include_all: true)` to test that *any* SQL query was made.

Reply:
IMO `assert_sql` means I want to assert some query matches something. 

If I only care about the count, it should be `assert_queries`.

Reply:
Yeah, I agree with you. There is a need to update quite a few test cases to specify these concrete sqls, like in https://github.com/rails/rails/pull/50373/files#diff-9bc4b7719b110cde3044f7e2ce6f8cda8f8e3263790d600e6d6792e40c02fa70R831

Will try to update with the suggestion.

Reply:
Also, does the count apply to the matching queries or all queries?

Reply:
Made matcher a required argument.